x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/view/ViewService.java

```diff
@@ -0,0 +1,321 @@
+    }
+
+    protected ViewMetadata getMetadata(ProjectMetadata projectMetadata) {
+        return projectMetadata.custom(ViewMetadata.TYPE, ViewMetadata.EMPTY);
+    }
+
+    ViewMetadata getMetadata(ProjectId projectId) {
+        return getMetadata(clusterService.state().metadata().getProject(projectId));
+    }
+
+    /**
+     * Adds or modifies a view by name.
+     */
+    public void putView(ProjectId projectId, PutViewAction.Request request, ActionListener<AcknowledgedResponse> listener) {
+        if (viewsFeatureEnabled() == false) {
+            listener.onFailure(new IllegalArgumentException("ESQL views are not enabled"));
+            return;
+        }
+
+        final View view = request.view();
+        final ProjectMetadata metadata = clusterService.state().metadata().getProject(projectId);
+        try {
+            validatePutView(metadata, view);
+        } catch (Exception e) {
+            listener.onFailure(e);
+            return;
+        }
+        final AckedClusterStateUpdateTask task = new AckedClusterStateUpdateTask(request, listener) {
+            @Override
+            public ClusterState execute(ClusterState currentState) {
+                final ProjectMetadata project = currentState.metadata().getProject(projectId);
+                final ViewMetadata viewMetadata = project.custom(ViewMetadata.TYPE, ViewMetadata.EMPTY);
+                final View currentView = viewMetadata.getView(view.name());
+                if (view.equals(currentView)) {
+                    // The update is a no-op, so no change is necessary
+                    return currentState;
+                }
+                // Validate the view again, because it could have become invalid between the pre-task submission and post-task submission.
+                validatePutView(metadata, view);
+                final Map<String, View> updatedViews = new HashMap<>(viewMetadata.views());
+                updatedViews.put(view.name(), view);
+                var metadata = ProjectMetadata.builder(project).putCustom(ViewMetadata.TYPE, new ViewMetadata(updatedViews));
+                return ClusterState.builder(currentState).putProjectMetadata(metadata).build();
+            }
+        };
+        taskQueue.submitTask("update-esql-view-metadata-[" + view.name() + "]", task, task.timeout());
+    }
+
+    /**
+     * Removes a view from the cluster state.
+     */
+    public void deleteView(ProjectId projectId, DeleteViewAction.Request request, ActionListener<AcknowledgedResponse> listener) {
+        if (viewsFeatureEnabled() == false) {
+            listener.onFailure(new IllegalArgumentException("ESQL views are not enabled"));
+            return;
+        }
+        final String name = request.name();
+        MetadataCreateIndexService.validateIndexOrAliasName(
+            name,
+            (viewName, error) -> new IllegalArgumentException("invalid view name [" + name + "], " + error)
+        );
```

Comment:
I don't think there's a lot of value in this check; is there? We just use the name for some map lookups and, fortunately, there's no such thing as Java-Map-injections. We do output the name in the failure message below; is that why you have this added check?

It feels a bit random here to have this check here, but not do the proper name validation we do in `PutViewAction` (i.e. the lowercase check).

Reply:
Yes I agree, I do not like having the duplicate validation (and it's already in the REST side anyway).